File: tests/test_exec_fix.py
"""
Unit tests for the exec vulnerability fix in vulnerabilities/exec/source/low.php

This test verifies that escapeshellarg() is properly applied to user input
to prevent shell injection attacks.
"""

import re
import os


def get_file_content():
    """Read the low.php file content"""
    file_path = os.path.join(
        os.path.dirname(__file__), "..", "vulnerabilities", "exec", "source", "low.php"
    )
    with open(file_path, "r") as f:
        return f.read()


def test_windows_branch_uses_escapeshellarg():
    """Test that the Windows branch uses escapeshellarg"""
    content = get_file_content()

    # Check for the pattern: 'ping  ' . escapeshellarg( $target )
    pattern = r"shell_exec\s*\(\s*['\"]ping\s+['\"]\s*\.\s*escapeshellarg\s*\(\s*\$target\s*\)"

    assert re.search(
        pattern, content
    ), "escapeshellarg not found in Windows branch for $target variable"


def test_unix_branch_uses_escapeshellarg():
    """Test that the *nix branch uses escapeshellarg"""
    content = get_file_content()

    # Check for the pattern: 'ping  -c 4 ' . escapeshellarg( $target )
    pattern = r"shell_exec\s*\(\s*['\"]ping\s+-c\s+4\s+['\"]\s*\.\s*escapeshellarg\s*\(\s*\$target\s*\)"

    assert re.search(
        pattern, content
    ), "escapeshellarg not found in *nix branch for $target variable"


def test_target_variable_not_used_unescaped():
    """Test that $target variable is not used unescaped in shell_exec"""
    content = get_file_content()

    lines = content.split("\n")

    for line_num, line in enumerate(lines, 1):
        if "shell_exec" in line and "$target" in line:
            # Check if this line has $target without escapeshellarg
            if not re.search(r"escapeshellarg\s*\(\s*\$target\s*\)", line):
                assert (
                    False
                ), f"Unescaped $target found in shell_exec on line {line_num}: {line.strip()}"


def test_static_command_parts_not_escaped():
    """Test that static command parts (ping, flags) are not escaped"""
    content = get_file_content()

    # Verify that 'ping' command itself is not escaped
    # We should NOT see escapeshellarg('ping') or escapeshellarg('ping -c 4')
    bad_patterns = [
        r"escapeshellarg\s*\(\s*['\"]ping",
        r"escapeshellarg\s*\(\s*['\"]ping\s+-c\s+4",
    ]

    for pattern in bad_patterns:
        assert not re.search(
            pattern, content
        ), f"Static command incorrectly escaped with pattern: {pattern}"


def test_command_structure_intact():
    """Test that the fix doesn't break the command structure"""
    content = get_file_content()

    # Verify the basic structure is maintained
    assert "shell_exec" in content, "shell_exec function not found"
    assert "ping" in content, "ping command not found"
    assert "-c 4" in content, "-c 4 flag not found for *nix"
    assert "$target" in content, "$target variable not found"


def test_both_branches_have_fix():
    """Test that both Windows and *nix branches have the fix applied"""
    content = get_file_content()

    # Count occurrences of escapeshellarg($target)
    pattern = r"escapeshellarg\s*\(\s*\$target\s*\)"
    matches = re.findall(pattern, content)

    # Should have exactly 2 occurrences (one for Windows, one for *nix)
    assert (
        len(matches) == 2
    ), f"Expected 2 occurrences of escapeshellarg($target), found {len(matches)}"


def test_no_double_escaping():
    """Test that there's no double escaping (escapeshellarg applied twice)"""
    content = get_file_content()

    # Look for nested escapeshellarg calls
    pattern = r"escapeshellarg\s*\(\s*escapeshellarg"

    assert not re.search(
        pattern, content
    ), "Double escaping detected (nested escapeshellarg calls)"


def test_original_variable_preserved():
    """Test that the original $target variable expression is preserved inside escapeshellarg"""
    content = get_file_content()

    # Verify that we're escaping exactly $target, not a modified version
    # Should be escapeshellarg( $target ), not escapeshellarg( trim($target) ) or similar
    pattern = r"escapeshellarg\s*\(\s*\$target\s*\)"
    matches = re.findall(pattern, content)

    assert (
        len(matches) >= 2
    ), "Original $target variable not preserved in escapeshellarg calls"


def test_concatenation_structure_preserved():
    """Test that the concatenation structure is preserved"""
    content = get_file_content()

    # Verify that concatenation with '.' is still used
    # Pattern: 'command' . escapeshellarg($target)
    windows_pattern = r"['\"]ping\s+['\"]\s*\.\s*escapeshellarg"
    unix_pattern = r"['\"]ping\s+-c\s+4\s+['\"]\s*\.\s*escapeshellarg"

    assert re.search(
        windows_pattern, content
    ), "Windows branch concatenation structure not preserved"
    assert re.search(
        unix_pattern, content
    ), "*nix branch concatenation structure not preserved"


def test_file_still_uses_request_input():
    """Test that the file still gets input from $_REQUEST"""
    content = get_file_content()

    # Verify that input is still retrieved from $_REQUEST['ip']
    pattern = r"\$target\s*=\s*\$_REQUEST\s*\[\s*['\"]ip['\"]\s*\]"

    assert re.search(
        pattern, content
    ), "$target assignment from $_REQUEST['ip'] not found"

File: tests/test_exec_fix.php
<?php
/**
 * Unit tests for the exec vulnerability fix in vulnerabilities/exec/source/low.php
 * 
 * This test verifies that escapeshellarg() is properly applied to user input
 * to prevent shell injection attacks.
 */

class ExecFixTest {
    private $testResults = [];
    private $testsPassed = 0;
    private $testsFailed = 0;

    /**
     * Test that escapeshellarg properly escapes malicious input
     */
    public function testEscapeshellargEscapesMaliciousInput() {
        $testName = "escapeshellarg escapes malicious input";
        
        // Test various shell injection payloads
        $maliciousInputs = [
            '127.0.0.1; cat /etc/passwd',
            '127.0.0.1 && whoami',
            '127.0.0.1 | ls -la',
            '127.0.0.1`id`',
            '127.0.0.1$(whoami)',
            "127.0.0.1\ncat /etc/passwd",
            '127.0.0.1; rm -rf /',
        ];

        $allPassed = true;
        foreach ($maliciousInputs as $input) {
            $escaped = escapeshellarg($input);
            
            // Verify the escaped string is wrapped in single quotes
            if (!preg_match("/^'.*'$/", $escaped)) {
                $allPassed = false;
                $this->recordTest($testName, false, "Input not properly quoted: $input");
                continue;
            }
            
            // Verify dangerous characters are neutralized
            // After escapeshellarg, the entire input becomes a single quoted string
            // so shell metacharacters lose their special meaning
            if (strpos($escaped, ';') !== false && strpos($escaped, "';'") === false) {
                $allPassed = false;
                $this->recordTest($testName, false, "Semicolon not properly escaped: $input");
                continue;
            }
        }

        if ($allPassed) {
            $this->recordTest($testName, true);
        }
    }

    /**
     * Test that the fix is applied in the Windows branch
     */
    public function testWindowsBranchUsesEscapeshellarg() {
        $testName = "Windows branch uses escapeshellarg";
        
        $fileContent = file_get_contents(__DIR__ . '/../vulnerabilities/exec/source/low.php');
        
        // Check for the pattern: 'ping  ' . escapeshellarg( $target )
        $windowsPattern = "/shell_exec\s*\(\s*['\"]ping\s+['\"]\\s*\\.\\s*escapeshellarg\s*\(\s*\\\$target\s*\)/";
        
        if (preg_match($windowsPattern, $fileContent)) {
            $this->recordTest($testName, true);
        } else {
            $this->recordTest($testName, false, "escapeshellarg not found in Windows branch");
        }
    }

    /**
     * Test that the fix is applied in the *nix branch
     */
    public function testUnixBranchUsesEscapeshellarg() {
        $testName = "*nix branch uses escapeshellarg";
        
        $fileContent = file_get_contents(__DIR__ . '/../vulnerabilities/exec/source/low.php');
        
        // Check for the pattern: 'ping  -c 4 ' . escapeshellarg( $target )
        $unixPattern = "/shell_exec\s*\(\s*['\"]ping\s+-c\s+4\s+['\"]\\s*\\.\\s*escapeshellarg\s*\(\s*\\\$target\s*\)/";
        
        if (preg_match($unixPattern, $fileContent)) {
            $this->recordTest($testName, true);
        } else {
            $this->recordTest($testName, false, "escapeshellarg not found in *nix branch");
        }
    }

    /**
     * Test that $target variable is not used unescaped
     */
    public function testTargetVariableNotUsedUnescaped() {
        $testName = "\$target variable not used unescaped in shell_exec";
        
        $fileContent = file_get_contents(__DIR__ . '/../vulnerabilities/exec/source/low.php');
        
        // Look for dangerous pattern: shell_exec with $target not wrapped in escapeshellarg
        // This regex looks for shell_exec containing $target but NOT preceded by escapeshellarg(
        $lines = explode("\n", $fileContent);
        $foundVulnerability = false;
        
        foreach ($lines as $lineNum => $line) {
            if (strpos($line, 'shell_exec') !== false && strpos($line, '$target') !== false) {
                // Check if this line has $target without escapeshellarg
                if (!preg_match('/escapeshellarg\s*\(\s*\$target\s*\)/', $line)) {
                    $foundVulnerability = true;
                    $this->recordTest($testName, false, "Unescaped \$target found on line " . ($lineNum + 1));
                    break;
                }
            }
        }
        
        if (!$foundVulnerability) {
            $this->recordTest($testName, true);
        }
    }

    /**
     * Test that static command parts are not escaped
     */
    public function testStaticCommandPartsNotEscaped() {
        $testName = "Static command parts not escaped";
        
        $fileContent = file_get_contents(__DIR__ . '/../vulnerabilities/exec/source/low.php');
        
        // Verify that 'ping' command itself is not escaped
        // We should NOT see escapeshellarg('ping') or escapeshellarg('ping -c 4')
        $badPatterns = [
            "/escapeshellarg\s*\(\s*['\"]ping/",
            "/escapeshellarg\s*\(\s*['\"]ping\s+-c\s+4/",
        ];
        
        $allGood = true;
        foreach ($badPatterns as $pattern) {
            if (preg_match($pattern, $fileContent)) {
                $allGood = false;
                $this->recordTest($testName, false, "Static command incorrectly escaped");
                break;
            }
        }
        
        if ($allGood) {
            $this->recordTest($testName, true);
        }
    }

    /**
     * Test that escapeshellarg preserves valid IP addresses
     */
    public function testEscapeshellargPreservesValidInput() {
        $testName = "escapeshellarg preserves valid input";
        
        $validInputs = [
            '127.0.0.1',
            '192.168.1.1',
            'localhost',
            'example.com',
        ];

        $allPassed = true;
        foreach ($validInputs as $input) {
            $escaped = escapeshellarg($input);
            
            // The escaped version should contain the original input
            // escapeshellarg wraps in single quotes, so we check the content
            $unquoted = trim($escaped, "'");
            
            if ($unquoted !== $input) {
                $allPassed = false;
                $this->recordTest($testName, false, "Valid input modified: $input -> $escaped");
                break;
            }
        }

        if ($allPassed) {
            $this->recordTest($testName, true);
        }
    }

    /**
     * Test that the fix doesn't break the command structure
     */
    public function testCommandStructureIntact() {
        $testName = "Command structure remains intact";
        
        $fileContent = file_get_contents(__DIR__ . '/../vulnerabilities/exec/source/low.php');
        
        // Verify the basic structure is maintained:
        // - shell_exec is still used
        // - ping command is still present
        // - -c 4 flag is still present for *nix
        $checks = [
            'shell_exec' => strpos($fileContent, 'shell_exec') !== false,
            'ping command' => strpos($fileContent, 'ping') !== false,
            '-c 4 flag' => strpos($fileContent, '-c 4') !== false,
        ];
        
        $allPassed = true;
        foreach ($checks as $checkName => $result) {
            if (!$result) {
                $allPassed = false;
                $this->recordTest($testName, false, "Missing: $checkName");
                break;
            }
        }
        
        if ($allPassed) {
            $this->recordTest($testName, true);
        }
    }

    /**
     * Record a test result
     */
    private function recordTest($testName, $passed, $message = '') {
        if ($passed) {
            $this->testsPassed++;
            $this->testResults[] = "✓ PASS: $testName";
        } else {
            $this->testsFailed++;
            $this->testResults[] = "✗ FAIL: $testName" . ($message ? " - $message" : "");
        }
    }

    /**
     * Run all tests
     */
    public function runAllTests() {
        echo "Running exec vulnerability fix tests...\n";
        echo str_repeat("=", 70) . "\n\n";

        $this->testEscapeshellargEscapesMaliciousInput();
        $this->testWindowsBranchUsesEscapeshellarg();
        $this->testUnixBranchUsesEscapeshellarg();
        $this->testTargetVariableNotUsedUnescaped();
        $this->testStaticCommandPartsNotEscaped();
        $this->testEscapeshellargPreservesValidInput();
        $this->testCommandStructureIntact();

        echo "\nTest Results:\n";
        echo str_repeat("-", 70) . "\n";
        foreach ($this->testResults as $result) {
            echo $result . "\n";
        }
        echo str_repeat("=", 70) . "\n";
        echo "Total: " . ($this->testsPassed + $this->testsFailed) . " tests\n";
        echo "Passed: " . $this->testsPassed . "\n";
        echo "Failed: " . $this->testsFailed . "\n";

        return $this->testsFailed === 0;
    }
}

// Run tests if executed directly
if (php_sapi_name() === 'cli') {
    $tester = new ExecFixTest();
    $success = $tester->runAllTests();
    exit($success ? 0 : 1);
}
?>
